design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,140 @@
+# Ceph COSI Driver Support
+
+## Targeted for v1.12
+
+## Background
+
+Container Object Storage Interface (COSI) is a specification for container orchestration frameworks to manage object storage. Even though there is no standard protocol defined for Object Store, it has flexibility to add support for all. The COSI spec abstracts common storage features such as create/delete buckets, grant/revoke access to buckets, attach/detach buckets, and more. COSI released v1alpha1 with Kubernetes 1.25. COSI is a new project and is not yet fully integrated by Kubernetes.
+More details about COSI can be found [here](https://kubernetes.io/blog/2022/09/02/cosi-kubernetes-object-storage-management/)
+It is projected that COSI will be the only supported object storage driver in the near future.
+
+## COSI Driver Deployment
+
+The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace. The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). The Ceph COSI driver can be deployed in any namespace not along with the COSI controller. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:
+
+## Integration plan with Rook
+
+The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There should be option in `operator.yaml` to validate whether the COSI controller exists and another option at object store level to bring up the Ceph COSI driver. Each Ceph object store should have its own COSI driver.
+
+### Pre-requisites
+
+- COSI CRDs should be installed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
+```
+
+- COSI controller should be deployed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
+```
+
+### End to End Workflow
+
+#### Changes in Common.yaml
+
+Following contents need to be append to `common.yaml` :
+
+- <https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml>
+- <https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml>
+
+#### cephcosi.ceph.rook.io CRD
```

Comment:
After discussion today:

COSI's intent is one driver per k8s cluster, so let's keep to that. 

Having a default behavior in Rook to deploy Ceph COSI driver if any CephObjectStore is deployed is a good one. Users shouldn't need to set any COSI config by default, just like with CSI today.

CRD still gives us some benefit in having a typed API that helps users with config and config validation compared to ConfigMap. Let's use CRD.

CRD can be deployed optionally if users want to override Rook's built-in defaults for COSI deployment. For example, for overriding the container image, labels, resource requests/limits, tolerations, etc. on Ceph COSI pods.

I had a thought that if we are using the CRD primarily for configuration rather than control of whether the thing exists, it might be good to call it `CephCOSIConfig` rather than `CephCOSI`. @travisn ? The CRD may also benefit from a `deploymentStrategy` option that has options `AlwaysDeploy`, `NeverDeploy`, and `AutoDeploy` where `AutoDeploy` is Rook's internal default to deploy when a CephObjectStore exists.

Reply:
> COSI's intent is one driver per k8s cluster, so let's keep to that.

Agreed, let's go with one driver per Rook operator, of which we expect one.
 
> Having a default behavior in Rook to deploy Ceph COSI driver if any CephObjectStore is deployed is a good one. Users shouldn't need to set any COSI config by default, just like with CSI today.
> 
> CRD still gives us some benefit in having a typed API that helps users with config and config validation compared to ConfigMap. Let's use CRD.

Agreed on the CRD since it's scoped with the operator, instead of with the CephCluster

> CRD can be deployed optionally if users want to override Rook's built-in defaults for COSI deployment. For example, for overriding the container image, labels, resource requests/limits, tolerations, etc. on Ceph COSI pods.

> I had a thought that if we are using the CRD primarily for configuration rather than control of whether the thing exists, it might be good to call it `CephCOSIConfig` rather than `CephCOSI`. @travisn ? 

IMO CRDs are implicitly about config, so adding the `Config` suffix seems unnecessary to me. What about `CephCOSIDriver`? 

> The CRD may also benefit from a `deploymentStrategy` option that has options `AlwaysDeploy`, `NeverDeploy`, and `AutoDeploy` where `AutoDeploy` is Rook's internal default to deploy when a CephObjectStore exists.

What is the difference between `AlwaysDeploy` and `AutoDeploy`? They seem the same to me at least in the case of the COSI driver. If we don't have a difference, I'd vote just for a boolean which simply allows disabling the driver.

Reply:
Auto --> deploys when first object store is created
Force --> will create when CRD is created, do not wait for object store
Disable --> disbales ceph cosi driver

Reply:
I see now, the main difference is `Force` allows creating the driver before the first object store is created.